### PR TITLE
Strict events

### LINKED ISSUE #314: Refactor event emitting
Currently typescript compiler doesn't know which event can be expected nor what data is sent.
We should use strictly typed events. Helper library: https://github.com/bterlson/strict-event-emitter-types
Blog post explainer: https://medium.com/@bterlson/strongly-typed-event-emitters-2c2345801de8

Tasks:
- create interface of possible events and their payload for each event emitter
- every class that's implementing EventEmitter should implement StrictEventEmitter.

### PR DESCRIPTION
**Pull request by mpetrunic**
- enables compile time check for supported events and payloads

resolves #314

### CODE REVIEW

File: packages/lodestar/src/network/interface.ts [hunk 1]
```diff
@@ -4,12 +4,22 @@
...
-import {RequestId, Method} from "../constants";
+import {Attestation, BeaconBlock, RequestBody, ResponseBody, Shard,} from "../types";
+import {ATTESTATION_TOPIC, BLOCK_TOPIC, Method, RequestId} from "../constants";
+import StrictEventEmitter from "strict-event-emitter-types";
 
-export interface INetwork extends EventEmitter {
+interface INetworkEvents {
+  [BLOCK_TOPIC]: (block: BeaconBlock) => void;
+  [ATTESTATION_TOPIC]: (attestation: Attestation) => void;
+  ["gossipsub:heartbeat"]: void;
```
Comment (GregTheGreek) [inline packages/lodestar/src/network/interface.ts:18]: Maybe we can make all of these constants

Gossip.heartbeat?

**Review by GregTheGreek — CHANGES_REQUESTED** [review]

File: packages/lodestar/src/network/interface.ts (hunk 1 above)
Comment (mpetrunic) [reply packages/lodestar/src/network/interface.ts:18]: I thinks thats how is defined in spec. Anyways, I didn't modify logic haha

File: packages/lodestar/src/network/libp2p/network.ts [hunk 2]
```diff
@@ -79,30 +77,32 @@ export class Libp2pNetwork extends EventEmitter implements INetwork {
...
     } catch (e) {
     }
-  }
-  private handleIncomingShardAttestation(msg: any): void {
+  };
+  private handleIncomingShardAttestation = (msg: any): void => {
     try {
       const attestation: Attestation = deserialize(msg.data, this.config.types.Attestation);
+      // @ts-ignore
+      // we cannot type hint this
```
Comment (wemeetagain) [inline packages/lodestar/src/network/libp2p/network.ts:100]: can we add a comment about this in INetworkEvents?

File: packages/lodestar/src/network/libp2p/rpc.ts [hunk 3]
```diff
@@ -152,6 +154,7 @@ export class NetworkRpc extends EventEmitter {
         delete this.requests[id];
         reject(new Error(`request timeout, method ${method}, id ${id}`));
       }, this.requestTimeout);
+      // @ts-ignore
```
Comment (wemeetagain) [inline packages/lodestar/src/network/libp2p/rpc.ts:164]: same, can we include a comment about `response ${id}` events in INetworkRpcEvents?

**Review by wemeetagain — CHANGES_REQUESTED** [review]
Looks good :)
Only comment about adding comments for the events that we aren't able to add typings for.

File: packages/lodestar/src/chain/chain.ts [hunk 4]
```diff
@@ -19,7 +19,7 @@ import {getEmptyBlock, initializeBeaconStateFromEth1, isValidGenesisState} from
 import {stateTransition} from "./stateTransition";
 import {LMDGHOST, StatefulDagLMDGHOST} from "./forkChoice";
 import {computeEpochOfSlot, getAttestingIndices} from "./stateTransition/util";
-import {IBeaconChain} from "./interface";
+import {ChainEventEmmitter, IBeaconChain} from "./interface";
```
Comment (wemeetagain) [inline packages/lodestar/src/chain/chain.ts]: One m in emitter

**Review by wemeetagain — CHANGES_REQUESTED** [review]

**Review by wemeetagain — APPROVED** [review]
:smiley_cat:

**Review by GregTheGreek — APPROVED** [review]
> And I think it's gonna be a long, long time
> Till touch down brings me round again to find
> I'm not the man they think I am at home
> Oh no no no, I'm a rocket man
> Rocket man, burning out his fuse up here alone
🚀 🚀 🚀 🚀 🚀 🚀 🚀 🚀 🚀 🚀 🚀 🚀 🚀 🚀

**Pull request merged**